From 0da8c6ac0e787f6b74c29f86046e5df2881567a6 Mon Sep 17 00:00:00 2001
From: Appana Durga Kedareswara Rao <appana.durga.rao@xilinx.com>
Date: Mon, 6 Jan 2025 22:44:49 -1200
Subject: [PATCH 084/110] net: axienet: Adding highmem support

The driver issue was observed on Microblaze with Higmem enabled and 1GB
memory.
Add highmem support for skb fragments to resolve it.
page_address() assumes to work with a lowmem page only but fragments
can be placed in highmem pages which has to be handled differently
then lomem pages.

This patch was inspired by:
"net: mv643xx_eth: Fix highmem support in non-TSO egress path"
(sha1: 9e91141)
where the highmem problem was observed.

skb_frag_dma_map() calls dma_map_page() that's why it is necessary to
distinguish if dma_map_simple() or dma_map_page() is called.
This is done by saving flag to every TX BD in tx_desc_mapping field.

There is also reference in Documentation/DMA-API-HOWTO.txt that
dma_map_single cannot be used with HIGMEM.
"Using CPU pointers like this for single mappings has a
disadvantage:
you cannot reference HIGHMEM memory in this way.  Thus, there is a
map/unmap interface pair akin to dma_{map,unmap}_single().  These
interfaces deal with page/offset pairs instead of CPU pointers."

Signed-off-by: Appana Durga Kedareswara Rao <appana.durga.rao@xilinx.com>
Acked-by: Harini Katakam <harini.katakam@amd.com>
Signed-off-by: Suraj Gupta <suraj.gupta2@amd.com>
Reviewed-by: Radhey Shyam Pandey <radhey.shyam.pandey@amd.com>
State: pending
Message-ID: <20250107104507.359136-12-suraj.gupta2@amd.com>
---
 drivers/net/ethernet/xilinx/xilinx_axienet.h  |  4 ++
 .../net/ethernet/xilinx/xilinx_axienet_main.c | 51 ++++++++++---------
 2 files changed, 31 insertions(+), 24 deletions(-)

diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet.h b/drivers/net/ethernet/xilinx/xilinx_axienet.h
index ca976955f386c..0c1ffa5f6d7a9 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet.h
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet.h
@@ -506,6 +506,7 @@ enum temac_stat {
  * @ptp_tx_ts_tag: Tag value of 2 step timestamping if timestamping is enabled
  *		   Otherwise reserved.
  * @tx_skb:	  Transmit skb address
+ * @tx_desc_mapping: Tx Descriptor DMA mapping type.
  */
 struct axidma_bd {
 	u32 next;	/* Physical address of next buffer descriptor */
@@ -525,9 +526,12 @@ struct axidma_bd {
 	phys_addr_t ptp_tx_skb;
 	u32 ptp_tx_ts_tag;
 	phys_addr_t tx_skb;
+	u32 tx_desc_mapping;
 } __aligned(XAXIDMA_BD_MINIMUM_ALIGNMENT);
 
 #define XAE_NUM_MISC_CLOCKS 3
+#define DESC_DMA_MAP_SINGLE 0
+#define DESC_DMA_MAP_PAGE 1
 
 /**
  * struct skbuf_dma_descriptor - skb for each dma descriptor
diff --git a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
index 840aef9cec05c..4c7d0fecb6160 100644
--- a/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
+++ b/drivers/net/ethernet/xilinx/xilinx_axienet_main.c
@@ -1024,12 +1024,18 @@ static int axienet_free_tx_chain(struct axienet_local *lp, u32 first_bd,
 		/* Ensure we see complete descriptor update */
 		dma_rmb();
 		phys = desc_get_phys_addr(lp, cur_p);
-		dma_unmap_single(lp->dev, phys,
-				 (cur_p->cntrl & XAXIDMA_BD_CTRL_LENGTH_MASK),
-				 DMA_TO_DEVICE);
-
-		if (cur_p->skb && (status & XAXIDMA_BD_STS_COMPLETE_MASK)) {
-			napi_consume_skb(cur_p->skb, budget);
+		if (cur_p->tx_desc_mapping == DESC_DMA_MAP_PAGE)
+			dma_unmap_page(lp->dev, phys,
+				       cur_p->cntrl &
+				       XAXIDMA_BD_CTRL_LENGTH_MASK,
+				       DMA_TO_DEVICE);
+		else
+			dma_unmap_single(lp->dev, phys,
+					 cur_p->cntrl &
+					 XAXIDMA_BD_CTRL_LENGTH_MASK,
+					 DMA_TO_DEVICE);
+		if (cur_p->tx_skb && (status & XAXIDMA_BD_STS_COMPLETE_MASK)) {
+			napi_consume_skb((struct sk_buff *)cur_p->tx_skb, budget);
 			packets++;
 		}
 
@@ -1040,6 +1046,8 @@ static int axienet_free_tx_chain(struct axienet_local *lp, u32 first_bd,
 		cur_p->status = 0;
 		cur_p->tx_skb = 0;
 
+		/* ensure our transmit path and device don't prematurely see status cleared */
+		wmb();
 		if (sizep)
 			*sizep += status & XAXIDMA_BD_STS_ACTUAL_LEN_MASK;
 	}
@@ -1313,7 +1321,6 @@ axienet_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 	u32 num_frag;
 	u32 csum_start_off;
 	u32 csum_index_off;
-	skb_frag_t *frag;
 	dma_addr_t tail_p, phys;
 	u32 orig_tail_ptr, new_tail_ptr;
 	struct axienet_local *lp = netdev_priv(ndev);
@@ -1450,27 +1457,23 @@ axienet_start_xmit(struct sk_buff *skb, struct net_device *ndev)
 	}
 	desc_set_phys_addr(lp, phys, cur_p);
 	cur_p->cntrl = (skb_headlen(skb) | XAXIDMA_BD_CTRL_TXSOF_MASK) + pad;
+	cur_p->tx_desc_mapping = DESC_DMA_MAP_SINGLE;
 
 	for (ii = 0; ii < num_frag; ii++) {
-		if (++new_tail_ptr >= lp->tx_bd_num)
-			new_tail_ptr = 0;
-		cur_p = &lp->tx_bd_v[new_tail_ptr];
+		u32 len;
+		skb_frag_t *frag;
+
+		if (++lp->tx_bd_tail >= lp->tx_bd_num)
+			lp->tx_bd_tail = 0;
+
+		cur_p = &lp->tx_bd_v[lp->tx_bd_tail];
 		frag = &skb_shinfo(skb)->frags[ii];
-		phys = dma_map_single(lp->dev,
-				      skb_frag_address(frag),
-				      skb_frag_size(frag),
-				      DMA_TO_DEVICE);
-		if (unlikely(dma_mapping_error(lp->dev, phys))) {
-			if (net_ratelimit())
-				netdev_err(ndev, "TX DMA mapping error\n");
-			ndev->stats.tx_dropped++;
-			axienet_free_tx_chain(lp, orig_tail_ptr, ii + 1,
-					      true, NULL, 0);
-			dev_kfree_skb_any(skb);
-			return NETDEV_TX_OK;
-		}
+		len = skb_frag_size(frag);
+		phys = skb_frag_dma_map(ndev->dev.parent, frag, 0, len,
+					DMA_TO_DEVICE);
 		desc_set_phys_addr(lp, phys, cur_p);
-		cur_p->cntrl = skb_frag_size(frag) + pad;
+		cur_p->cntrl = len + pad;
+		cur_p->tx_desc_mapping = DESC_DMA_MAP_PAGE;
 	}
 
 	cur_p->cntrl |= XAXIDMA_BD_CTRL_TXEOF_MASK;
-- 
2.43.0

